File: 模拟/处理/查找和替换模式/查找和替换模式.cpp
// https://leetcode.cn/problems/find-and-replace-pattern/

class Solution {
public:
    void process(string& s)
    {
        int cot = 1;
        unordered_map<char, int> ind;
        for (int i=0; i<s.size(); i++)
        {
            if (ind[s[i]] == 0)
                ind[s[i]] = cot ++;
            s[i] = ind[s[i]];
        }
    }

    vector<string> findAndReplacePattern(vector<string>& words, string pattern) {
        vector<string> ans;
        process(pattern);
        for (string s : words)
        {
            string str = s;
            process(str);
            if (str == pattern)
                ans.emplace_back(s);
        }
        return ans;
    }
};

File: 模拟/处理/查找和替换模式/查找和替换模式.py
# https://leetcode.cn/problems/find-and-replace-pattern/

class Solution:
    def process(self, s):
        cot, num, res = 0, {}, [i for i in s]
        for i in range(len(res)):
            if res[i] not in num.keys():
                num[res[i]] = cot
                cot += 1
            res[i] = num[res[i]]
        s = ''
        for i in res:
            s += str(i)
        return s
    def findAndReplacePattern(self, words: List[str], pattern: str) -> List[str]:
        return list(filter(lambda s: self.process(s)==self.process(pattern), words))